### PR TITLE
[FIX] maskers truncating cleaned signal to source image's dtype

### LINKED ISSUE #6525: [BUG] maskers transform() silently truncates cleaned signal to the source image's integer dtype
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating system

- [x] Linux

### Operating system version

Linux Ubuntu (kernel 7.0.0-29-generic)

### Python version

- [x] 3.10

### nilearn version

Dev version (main / 0.14.1.dev, also reproduces on `main` as of commit `7c73c85d9a49c0e60917099e8ace1f41c2dee9c7` and later). Does **not** reproduce on 0.13.1.

### Expected behavior

When `NiftiMasker(..., standardize="zscore_sample")` (or `"psc"`, or `standardize=True`) is used to `transform()` an image whose header dtype is an integer type (e.g. `int16`, common for raw/unprocessed BOLD data), the returned 2D signal array should keep the floating-point precision produced by standardization.

### Current behavior & error messages

`NiftiMasker.transform()` silently casts its output back to the **input image's own header dtype** whenever `NiftiMasker(dtype=...)` is left at its default (`None`). For an `int16` source image, this truncates the continuous z-scored/PSC signal down to a handful of integer levels (e.g. `{-2, -1, 0, 1, 2}`), destroying almost all of the signal's precision. No warning is raised.

This was introduced in #5511 ("[FIX] ensure image estimator can fit and preserve several dtype"), in `nilearn/maskers/nifti_masker.py::NiftiMasker.transform_single_imgs`:

```python
target_dtype = get_target_dtype(img_data_dtype(imgs), self.dtype)
if target_dtype is None:
    target_dtype = img_data_dtype(imgs)

return data.astype(target_dtype)
```

`get_target_dtype` returns `None` specifically to mean "no conversion needed" (per its own docstring: *"If None, no conversion is required"*), but the `if target_dtype is None:` branch overrides that `None` with the original image's raw dtype and casts unconditionally. This is presumably meant to mirror the `dtype`-preservation logic added for `inverse_transform` (`BaseMasker._set_inverse_transform_output_dtype`) elsewhere in the same PR, but applying it to the forward `transform()` path is incorrect: `transform()` output for a masker with `standardize` enabled is fundamentally a floating point quantity and should not be coerced back into the source image's discrete dtype.

I found this while comparing the output of `examples/02_decoding/plot_haxby_full_analysis.py` between the 0.13.1 release and current `main`: decoding AUC scores dropped noticeably across the board (e.g. mask_vt/chair: 0.93 → 0.78) purely because of this silent precision loss feeding into the SVM classifier, not because of any change to the example or the classifier itself. I bisected the regression down to commit `7c73c85d9a49c0e60917099e8ace1f41c2dee9c7` (#5511) using `git bisect` with a fixed dependency environment, and confirmed dependency-version differences and the unrelated `svc_l1`/`LinearSVC(dual=...)` change in the same time range are not responsible.

### Steps and code to reproduce bug

```python
import numpy as np
import nibabel as nib

from nilearn.maskers import NiftiMasker

rng = np.random.default_rng(0)
data = (rng.standard_normal((10, 10, 10, 20)) * 1000).astype(np.int16)
img = nib.Nifti1Image(data, affine=np.eye(4))

mask_data = np.zeros((10, 10, 10), dtype=np.int8)
mask_data[3:7, 3:7, 3:7] = 1
mask_img = nib.Nifti1Image(mask_data, affine=np.eye(4))

masker = NiftiMasker(mask_img=mask_img, standardize="zscore_sample")
X = masker.fit_transform(img)

print("dtype:", X.dtype)  # int16 -- should be float
print("n unique values in column 0:", len(np.unique(X[:, 0])))  # only 2-5 -- should be close to 20 (n_samples)
print(X[:, 0])
```

Output on current `main`:

```
dtype: int16
n unique values in column 0: 3
[ 0  0  0  1 -1  0 -1  0  0  1 -1 -1  0  0  1  0  0  0  1 -1]
```

Expected (and what 0.13.1 produces): `dtype: float32` (or `float64`) with ~20 distinct z-scored values.

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #6525

Use of AI tools:
- [x] LLM tools were used to generate at least part of the content of this pull-request.
- [ ] No LLM tools were used to generate any part of the content of this pull-request.

Changes proposed in this pull request:

- Fix `NiftiMasker`, `NiftiLabelsMasker`, `NiftiMapsMasker`, `NiftiSpheresMasker`, `SurfaceMasker`, `SurfaceLabelsMasker`, `SurfaceMapsMasker`, and `MultiSurfaceMasker` silently casting their `transform()` output back to the source image's own dtype whenever the masker's own `dtype` parameter was left at its default (`None`). For images with an integer header dtype (common for raw, unprocessed BOLD data), this truncated standardized (`"zscore_sample"`/`"psc"`) floating point signal down to a handful of integer levels, destroying almost all of its precision — with no warning.
- Restore the pre-regression behavior (introduced in #5511): when `dtype` is `None`, keep whatever dtype the extraction/cleaning pipeline actually produced instead of coercing it back to the source image's dtype. When `dtype` is explicitly set (including when it already happens to match the source dtype), keep casting to it, since intermediate computations (standardization, `lstsq`, ...) can still promote the working array to a different dtype.
- Fix `check_img_estimator_dtypes_transform` in `nilearn/_utils/estimator_checks.py`, which encoded the same incorrect assumption (added in the same #5511), and add a dedicated regression check, `check_masker_standardize_dtype`, registered for all maskers.
- Add a changelog entry.

I found this while comparing decoding results between the 0.13.1 release and current `main` for `examples/02_decoding/plot_haxby_full_analysis.py`: AUC scores dropped noticeably across the board purely because of this silent precision loss feeding into the classifier. I bisected the regression down to #5511 with a fixed dependency environment (ruling out dependency-version drift and an unrelated `svc_l1` change in the same range), then traced it to this dtype-casting logic. Full test suites for `nilearn/maskers`, `nilearn/decoding`, `nilearn/regions`, and `nilearn/_utils/tests/test_estimator_checks.py` pass (3220 passed, 0 failed).

### CODE REVIEW

File: examples/02_decoding/plot_haxby_full_analysis.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 ROI-based decoding analysis in Haxby et al. dataset
 ===================================================
 
-In this script we reproduce the data analysis
+In this script we reproduce the data  analysis
```
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_full_analysis.py]:
```suggestion
In this script we reproduce the data analysis
```
remove before merge: just there to trigger example build

**Review by bthirion — APPROVED** [review]
The doc build did not regenerate the examples, but LGTM.

File: nilearn/maskers/nifti_labels_masker.py [hunk 2]
```diff
@@ -812,11 +811,18 @@ def transform_single_imgs(self, imgs, confounds=None, sample_mask=None):
...
-        return region_signals.astype(target_dtype)
+        # target_dtype is None: no explicit dtype was requested,
+        # so keep the dtype produced by the extraction/cleaning pipeline
+        # (e.g. float after standardize)
+        # instead of forcing it back to the source image's dtype.
+        return (
+            region_signals
+            if target_dtype is None
+            else region_signals.astype(target_dtype)
+        )
```
Comment (hndgzkn) [inline nilearn/maskers/nifti_labels_masker.py]: This code seems to be repeating in all estimators. Instead of extracting `_get_target_dtype` out in `_BaseMasker`, I would suggest to rename it to contain this piece of code as well.

**Review by hndgzkn — APPROVED** [review]
LGTM a part from a suggestion.

File: nilearn/maskers/nifti_labels_masker.py (hunk 2 above)
Comment (Remi-Gau) [reply nilearn/maskers/nifti_labels_masker.py]: let me check if I can refactor further

File: examples/02_decoding/plot_haxby_full_analysis.py
Comment (Remi-Gau) [inline examples/02_decoding/plot_haxby_full_analysis.py:1]: touched the examples listed [here](https://github.com/nilearn/nilearn/issues/6524) to see if this PR gets us the performance we had in 0.13.1

File: nilearn/maskers/nifti_labels_masker.py (hunk 2 above)
Comment (Remi-Gau) [reply nilearn/maskers/nifti_labels_masker.py]: @hndgzkn 
I have done some further refactoring but I am not sure it was the one you meant.

I think that if you meant that the ternary expression of the return should also be in _get_target_dtype, then that's not going to work because sometimes _get_target_dtype happens very far from the return statement of the method.

maybe check out the branch directly to look at the code to make sure I am not missing anything.

**Comment by Remi-Gau** [comment]
||PR|main|0.13.1|
|-|-|-|-|
| 02_decoding/plot_haxby_searchlight.py | image | image | image |
| 02_decoding/plot_haxby_full_analysis.py | image | image | image |
| 02_decoding/plot_haxby_multiclass.py | image | image | image |
| 07_advanced/plot_advanced_decoding_scikit.py | SVC accuracy (tuned parameters): 0.858 | SVC accuracy (tuned parameters): 0.799 | SVC accuracy (tuned parameters): 0.858 |

- This brings back the results to their version 0.13.1 level so this closes #6524

File: nilearn/maskers/nifti_labels_masker.py (hunk 2 above)
Comment (hndgzkn) [reply nilearn/maskers/nifti_labels_masker.py]: > I think that if you meant that the ternary expression of the return should also be in _get_target_dtype, then that's not going to work because sometimes _get_target_dtype happens very far from the return statement of the method.

Yes, I meant that one as well, otherwise what you did is exactly what I was thinking of.

I see in NiftiMapsMasker that `_get_target_dtype` is called earlier, but it is not used until the ternary expression. Is it because `imgs_` might change that it is called earlier?

Comment (Remi-Gau) [reply nilearn/maskers/nifti_labels_masker.py]: not sure anymore: let me move later in the method to see if anything breaks
